Revamp WebCodecs documentation (#43697)
mdn/content: files/en-us/web/api/webcodecs_api/codec_selection/index.md

This guide explains how codec strings work and how to choose the right codec for common use cases.

## Decoding vs encoding

When **decoding** a video file, the codec is determined by how the file was originally encoded — you do not choose it. Demuxing libraries such as [MediaBunny](https://mediabunny.dev/) and [web-demuxer](https://github.com/bilibili/web-demuxer) will extract the correct codec string for a given file, which you can supply directly to {{domxref("VideoDecoder")}} during configuration.

When **encoding**, you choose the codec. The rest of this guide covers how to choose a codec.

## Choosing a codec family

Before choosing a codec string, like `vp09.00.40.08.00` or `avc1.4d0034`, it is worth reviewing the codec families.

### H.264 (AVC)

H.264 is one of the most widely supported codecs across browsers, operating systems, and consumer devices. It is the most common codec used in MP4 files, and applications which encode videos intended for playback in 3rd party software typically choose H.264 as a pragmatic choice for maximum compatibility.

> While popular, it is worth noting that H.264 is a patented codec. While browser vendors hold licenses covering the H.264 encoder implementations used by WebCodecs, the codec is subject to royalties in certain circumstances, and developers should be aware of the [Via Licensing Alliance AVC license](https://via-la.com/licensing-programs/avc-h-264/).

Reviewer: I'd drop the direct link to the program and just leave this as "circumstances. Developers should review usage with legal counsel."

> ---

Reviewer 2: @sb2702 
1. This is great. I haven't read the other docs yet and I have other tasks I need to do today, but there are some comments which I hope you find useful. 
2. I have a flavour of some of the options, but I'm still not certain of the general guidance. Is there a rule of thumb for someone writing a website - I mean I'd guess using AVC with the codec for the max resolution I want would be a safe choice. Is there any other guidance for use cases.

   Or to put it other way, there is enough information here to be able to work out your options, but it would be nice to give more concrete advice for a new starter wanting to address a couple of the most common codec use cases.

   Does that make sense?
3. Hope to start on the other docs next week.

Author: Awesome! 1. No worries, 2. I've added a quick "common use cases section" which should provide the "rule of thumb" guidance.  3. No worries, whenever you get some time!

title: Codec selection
slug: Web/API/WebCodecs_API/Codec_selection
page-type: guide
---

{{DefaultAPISidebar("WebCodecs API")}}

> The WebCodecs API requires working with fully specified codec strings, such as `vp09.00.40.08.00`, instead of ambiguous strings like `vp9` or `h264`. Fully specified codec strings detail not just the codec family but also the profile, level, and other parameters.

Reviewer 2: ```suggestion
The [WebCodecs API](/en-US/docs/Web/API/WebCodecs_API) requires working with fully specified codec strings, such as `vp09.00.40.08.00`, instead of ambiguous strings like `vp9` or `h264`. Fully specified codec strings detail not just the codec family but also the profile, level, and other parameters.
```

Reviewer 2: Just to be clear, when you say "requires" here, do you mean that the API expects a fully qualified string in the standardized format to set the encoding - if you enter a malformed string or something that isn't supported it will error in some way?

Reviewer 2: I might start this off with something like this, which has the same info, but leads in more softly and introduces your selection criteria up front.

> While developers commonly refer to codecs by their code identifier string, such as `vp9` or `h264`, there are many configuration profiles, levels, and other parameters that control exactly how the data is encoded and decoded.
> The [WebCodecs API](/en-US/docs/Web/API/WebCodecs_API) requires that codes are provided in a fully specified codec string, such as `vp09.00.40.08.00`, that provides all the relevant information.
>
> There are a many possible combinations of configuration parameters, and hence codec strings.
> Selecting the correct string depends on your use case, but is primarily influenced by compatibility concerns, and the hardware and software on which you want to run. [Off the top of my head!]
>
> This guide blah blah ...

Author: @hamishwillee, if you use a malformed codec string, you do get an error at encode time, when using an unsupported config and running `encoder.encode()`

`encoder.configure({
  codec: "vp9", // Won't throw an error at configure time, but will throw an error and encode timem 
  width: 1280,
  height: 720,
  bitrate: 1_000_000, // 1 Mbps
  framerate: 25,
});
`

I think the advice should be covered by the recommendation to use 

`
const { supported } = await VideoEncoder.isConfigSupported({
  codec: "avc1.4d0034",
  width: 1920,
  height: 1080,
});
`
That checks for unsupported configs.

In either case I've updated the language to hopefully be more beginner friendly

> This guide explains how codec strings work and how to choose the right codec for common use cases.

Reviewer 2: And perhaps given your excellent section at the end https://pr43697.review.mdn.allizom.net/en-US/docs/Web/API/WebCodecs_API/Codec_selection#checking_support_at_runtime you might want to do

```suggestion
This guide explains how codec strings work, how to choose the right codecs for common use cases, and common approaches for gracefully falling back to alternative codec strings when your preferences are unavailable.
```

Author: Done!

## Decoding vs encoding

> When **decoding** a video file, the codec is determined by how the file was originally encoded — you do not choose it. Demuxing libraries such as [MediaBunny](https://mediabunny.dev/) and [web-demuxer](https://github.com/bilibili/web-demuxer) will extract the correct codec string for a given file, which you can supply directly to {{domxref("VideoDecoder")}} during configuration.

Reviewer 2: FMI how are is the codec encoded in the file/stream? I.e. how do the libraries work out what string you're using?

Author: To be honest, I have no idea, this is definitely a question for @Vanilagy (author of MediaBunny) though I would guess the logic sits in files like[ this](https://github.com/Vanilagy/mediabunny/blob/main/src/isobmff/isobmff-demuxer.ts) and [this](https://github.com/Vanilagy/mediabunny/blob/main/src/matroska/matroska-demuxer.ts).

I briefly had my own shot at building a demuxer not because I wanted a demuxer, I just wanted to know how it worked under the hood and it's complex to say the least. I did put some info on that here:
https://webcodecsfundamentals.org/basics/muxing/#manual-demuxer

Reviewer 3: Extracting the codec string from a media file is non-trivial and differs for every container and codec. Some codecs only have one codec string (like flac), so that's easy. For AVC, HEVC, AV1 etc, there's either a "decoder metadata" section somewhere in the container for you (like AVCDecoderConfigurationRecord for ISOBMFF), or if that's not available (like for MPEG-TS files for example), you have no choice but to parse the actual bitstream of the encoded video frames.

In the case of AVC and HEVC, the stuff included in the codec string are things like profile, level, bit depth, chroma subsampling information, etc. This information is contained within the SPS/PPS/VPS NALUs (Network Abstraction Layer Units) which you'll need to locate within the bitstream and parse. These are sent in-band (within the bitstream) but only sparsely, typically at the start of the stream or with a key frame. They don't need to be repeated often since they don't change.

Anyway, it all gets quite crazy, and this is another reason why the WebCodecs API looks deceivingly simple on the surface but isn't in practice. These two files can give you an idea:

https://github.com/Vanilagy/mediabunny/blob/main/src/codec.ts
https://github.com/Vanilagy/mediabunny/blob/main/src/codec-data.ts

When **encoding**, you choose the codec. The rest of this guide covers how to choose a codec.

> ## Choosing a codec family

Reviewer 2: I was going to say that this info duplicates other info later. But I like it, and it is helpful in this context. Ignore me :-)

Reviewer 2: But there is a structural inconsistency with the audio section below - which mentions it is audio.

This is one possibility

```suggestion
## Choosing a video codec family
```

But perhaps you can have this just like (my preference, and make audio heading match)

```suggestion
## Video codecs

### Codec families
```

You will have to push the other headings down. THen have symmetry in the audio codecs

Before choosing a codec string, like `vp09.00.40.08.00` or `avc1.4d0034`, it is worth reviewing the codec families.

### H.264 (AVC)

H.264 is one of the most widely supported codecs across browsers, operating systems, and consumer devices. It is the most common codec used in MP4 files, and applications which encode videos intended for playback in 3rd party software typically choose H.264 as a pragmatic choice for maximum compatibility.

> While popular, it is worth noting that H.264 is a patented codec. While browser vendors hold licenses covering the H.264 encoder implementations used by WebCodecs, the codec is subject to royalties in certain circumstances. Developers should review usage with legal counsel.

Reviewer 2: Just FMI how concerning is this? I mean if my browser supports a codec via the underlying platform, and I write a web page that uses it, what is my risk? When is there a risk - just if I write a native app and import my own coded library from somewhere?

Author: Video codec patents really are a whole thing, and there's real ambiguity on legality and how much h264 patent claims are actually enforced. The technically correct answer to "what is my risk?" is, I don't know.

From a practical independent developer perspective, there is little practical risk on just choosing h264 and getting on with your life. For a generic tutorial on a more casual website, I'd feel more free to say something like "don't worry about it, practical risk is close to zero" but for MDN specifically as the authoritative resource, I think it'd be safer to be accurate and correct, even at the cost of some practicality. The text as is was my attempt at threading the needle of being correct / accurate and neutral on what is otherwise a sensitive topic in the video codec space.

Reviewer 2: I really hate being overly cautious in these case - it makes something fearsome when it shouldn't be. My understanding is that web app licencing is free if you're not making money from it yourself, and in a browser-app is paid for by the browser vendor. There's a tiered streaming model that does cost, but doesn't start until you're earning significant income from the video. 

I'll see if I can get advice from MDN team and get back to you. Something liike

```suggestion
The H.264 is a patented codec and is subject to royalties in certain circumstances. 
Generally you won't need to worry as a web application developer, because browser vendors already pay licenses covering the H.264 encoder implementations used by WebCodecs.
```

### HEVC (H.265)

HEVC offers better compression than H.264 but has significant gaps in browser encoding support outside of Apple platforms.
It is not recommended as a general-purpose encoding target.

Like H.264, HEVC is a patented codec. The codec is subject to royalties in certain circumstances. Developers should review usage with legal counsel.

## Codec-container compatibility

Not all codecs are supported by all containers.
The following table covers the two most common web video containers:

| Codec | MP4     | WebM |
| ----- | ------- | ---- |
| H.264 | Yes     | No   |
| VP9   | Partial | Yes  |
| AV1   | Partial | Yes  |
| HEVC  | Yes     | No   |

H.264 is the standard codec for MP4. VP9 and AV1 are the standard codecs for WebM.
While VP9 and AV1 have partial MP4 support in some environments, pairing them with WebM is more reliable.

> ## Choosing a codec string

Reviewer 2: Maybe. Just to shake it up a bit since we say "choosing" a few times.

```suggestion
## Codec string selection
```

Author: done!

For each codec family, there are hundreds of possible codec strings. The following tables provide a practical starting point for codec strings that maximize encoding compatibility.

> 

Reviewer 2: 1. What profile and level in the tables below. Are they relevant to selection? If so we should explain why
2. The support is shown as a %, which is nice. Is that of "all devices", "all browsers", "all OS"? Essentially I'm asking if, say, `98.9` means that 2 % of what will not work with if it gets sent a stream encoded with this string? Or to put it another way, why wouldn't I choose `vp09.00.40.08.00` which seems to have almost complete coverage at 2K

Author: I added a few lines about levels and profile in that section before the tables, hopefully that helps.

#2 - the support numbers are traffic weighted, like the numbers from caniuse.com, concretely from unique user session data from this dataset: https://zenodo.org/records/19187467. It should be fairly representative of general internet use for any given application.

Re:  "why wouldn't you use `vp09.00.40.08.00`, you should, that's why it's on the shorthand list. The main reason to not choose it is if you wanted mp4 outputs with h264, which has better playback support on native players, which was discussed earlier/higher up in the codec family selection:

"VP9 within WebM containers is widely supported across modern browsers, with coverage comparable to or exceeding H.264.

VP9 within WebM containers is also supported by native video players on Windows (Windows Media Player) and third-party players such as VLC, but currently lacks native playback support on macOS and iOS."

| `av01.0.05M.08` | 3.1   | 720p           | [87.9%](https://webcodecsfundamentals.org/codecs/av01.0.05M.08.html) |
| `av01.0.08M.08` | 4.0   | 1080p          | [87.8%](https://webcodecsfundamentals.org/codecs/av01.0.08M.08.html) |
| `av01.0.12M.08` | 5.0   | 4K             | [87.8%](https://webcodecsfundamentals.org/codecs/av01.0.12M.08.html) |

### HEVC

| Codec string       | Level | Max resolution | Support                                                                 |
| ------------------ | ----- | -------------- | ----------------------------------------------------------------------- |
| `hvc1.1.6.L90.B0`  | 3.0   | 1080p          | [73.6%](https://webcodecsfundamentals.org/codecs/hvc1.1.6.L90.B0.html)  |
| `hvc1.1.6.L150.B0` | 5.0   | 4K             | [73.6%](https://webcodecsfundamentals.org/codecs/hvc1.1.6.L150.B0.html) |
| `hvc1.1.6.L180.B0` | 6.0   | 8K             | [73.1%](https://webcodecsfundamentals.org/codecs/hvc1.1.6.L180.B0.html) |

See the [Codec Support Table](https://webcodecsfundamentals.org/datasets/codec-support-table/) for an exhaustive list of potential codec strings, and support across browsers & devices.

## Codec string format

> The WebCodecs API requires a fully qualified codec string rather than a general codec name. The string encodes the codec family, profile, level, and other parameters that affect which hardware can encode or decode the stream and at what resolution and quality.

Reviewer 2: You don't need to say this again.

```suggestion
The fully qualified codec string encodes the codec family, profile, level, and other parameters that affect which hardware can encode or decode the stream and at what resolution and quality.
```

The format for these codec strings is specified in the [W3C codec registry](https://www.w3.org/TR/webcodecs-codec-registry/), and the format is different for each codec family.

### H.264

`avc1.4d0034`

- `avc1` — H.264/AVC codec identifier
- `4d` — Profile IDC in hexadecimal (`4d` = Main profile)
- `00` — Constraint flags
- `34` — Level IDC in hexadecimal (`34` = level 5.2, supports up to 4K)

### VP9

`vp09.00.40.08.00`

- `vp09` — VP9 codec identifier
- `00` — Profile
- `40` — Level (`40` = level 4.0, supports up to 2K)
- `08` — Bit depth (8-bit)
- `00` — Chroma subsampling

### AV1

`av01.0.05M.08`

- `av01` — AV1 codec identifier
- `0` — Profile (Main)
- `05M` — Level and tier (`05` = level 3.1, `M` = Main tier)
- `08` — Bit depth (8-bit)

### HEVC

`hvc1.1.6.L150.B0`

- `hvc1` — HEVC codec identifier (MP4/QuickTime variant)
- `1` — Profile (`1` = Main profile)
- `6` — Compatibility flags
- `L150` — Level × 30 (`L150` = level 5.0, supports up to 4K)
- `B0` — Tier and constraint flags (`B0` = Main tier)

> ## Choosing an audio codec

Reviewer 2: As above, symmetry with the video section is good.

Author: I edited the headers to be symmetric with audio

### Opus

Opus is an open source codec with broad encoding support across browsers and platforms. It is the standard audio codec for WebM files, and the recommended choice for most WebCodecs audio encoding use cases.

### AAC

AAC is the standard audio codec for MP4 files and is required when targeting MP4 output. However, AAC encoding support in WebCodecs has notable gaps: it is not supported in Firefox on any platform, or in any browser on desktop Linux.

> AAC encoding is universally supported on Safari versions which support {{domxref("AudioEncoder")}} (Safari 26+), but previous versions of Safari do not support audio encoding in general.

Reviewer 2: ```suggestion
AAC encoding is universally supported on Safari versions that support {{domxref("AudioEncoder")}} (Safari 26+), but previous versions of Safari do not support audio encoding in general.
```

### MP3 and PCM

> MP3 and PCM are not supported as encoding targets in WebCodecs. PCM (uncompressed audio) is available as a {{domxref("AudioData")}} format for raw audio processing, but cannot be encoded via `AudioEncoder`.

Reviewer 2: WebCodecs? Either the API or the specific interface.

Author: Actually this is highly ambiguous, I updated this to reflect that it's a browser support issue and browser support is limited

---
title: Codec selection
slug: Web/API/WebCodecs_API/Codec_selection
page-type: guide
---

{{DefaultAPISidebar("WebCodecs API")}}

While developers commonly refer to codecs by their code identifier string, such as `vp9` or `h264`, there are many configuration profiles, levels, and other parameters that control exactly how the data is encoded and decoded.

The [WebCodecs API](/en-US/docs/Web/API/WebCodecs_API) requires working with fully specified codec strings, such as `vp09.00.40.08.00`, instead of ambiguous strings like `vp9` or `h264`. Fully specified codec strings detail not just the codec family but also the profile, level, and other parameters.

> Selecting the correct string depends on your use case, but is primarily influenced by compatibility concerns, and the hardware and software on which you want to run. This guide explains how codec strings work, how to choose the right codecs for common use cases, and common approaches for gracefully falling back to alternative codec strings when your preferences are unavailable.

Reviewer 2: ```suggestion
Selecting the correct string depends on your use case, but is primarily influenced by compatibility concerns, and the hardware and software on which you want to run. This guide explains how codec strings work, how to choose the right codecs for [common use cases](#common_use_cases), and common approaches for gracefully falling back to alternative codec strings when your preferences are unavailable.
```

> ## Decoding vs encoding

Reviewer 2: Is the this specific to video? If so, I would put it in the video section.

Author: I added some language to make sure it's not video specific

When **decoding** a video file, the codec is determined by how the file was originally encoded — you do not choose it. Demuxing libraries such as [Mediabunny](https://mediabunny.dev/) and [web-demuxer](https://github.com/bilibili/web-demuxer) will extract the correct codec string for a given file, which you can supply directly to {{domxref("VideoDecoder")}} during configuration.

When **encoding**, you choose the codec. The rest of this guide covers how to choose a codec.

> ## Choosing a video codec family

Reviewer 2: If it were me I would do this.  As it is now you group coded families at same level of as codec container compatibility.

````
## Video codecs

### Video codec families

#### H.264 (AVC)
...
### Codec-container compatibility  -- Note, one level up compared to the 

```

Author: Done!

The following tables provide a practical starting point for codec strings, with levels and profiles that maximize encoding compatibility.

#### H.264

| Codec string  | Profile  | Max resolution | Support                                                            |
| ------------- | -------- | -------------- | ------------------------------------------------------------------ |
| `avc1.42001f` | Baseline | 720p           | [99.6%](https://webcodecsfundamentals.org/codecs/avc1.42001f.html) |
| `avc1.4d0034` | Main     | 4K             | [98.9%](https://webcodecsfundamentals.org/codecs/avc1.4d0034.html) |
| `avc1.42003e` | Baseline | 8K             | [86.8%](https://webcodecsfundamentals.org/codecs/avc1.42003e.html) |
| `avc1.64003e` | High     | 8K             | [85.9%](https://webcodecsfundamentals.org/codecs/avc1.64003e.html) |

#### VP9

| Codec string       | Level | Max resolution | Support                                                                  |
| ------------------ | ----- | -------------- | ------------------------------------------------------------------------ |

> | `vp09.00.10.08.00` | 1     | 720p           | [99.98%](https://webcodecsfundamentals.org/codecs/vp09.00.10.08.00.html) |

Reviewer 4: This is factually incorrect. VP9 Level 1.0 has a tiny maximum picture size, only supporting up to 256×144 (144p) at 15fps. To support 720p, you need at least VP9 Level 3.0

Author: I switched the example to the right level (level 3)

| `vp09.00.50.08.00` | 5     | 4K             | [99.97%](https://webcodecsfundamentals.org/codecs/vp09.00.50.08.00.html) |
| `vp09.00.61.08.00` | 6.1   | 8K             | [99.97%](https://webcodecsfundamentals.org/codecs/vp09.00.61.08.00.html) |

#### AV1

| Codec string    | Level | Max resolution | Support                                                              |
| --------------- | ----- | -------------- | -------------------------------------------------------------------- |
| `av01.0.05M.08` | 3.1   | 720p           | [87.9%](https://webcodecsfundamentals.org/codecs/av01.0.05M.08.html) |
| `av01.0.08M.08` | 4.0   | 1080p          | [87.8%](https://webcodecsfundamentals.org/codecs/av01.0.08M.08.html) |
| `av01.0.12M.08` | 5.0   | 4K             | [87.8%](https://webcodecsfundamentals.org/codecs/av01.0.12M.08.html) |

#### HEVC

| Codec string       | Level | Max resolution | Support                                                                 |
| ------------------ | ----- | -------------- | ----------------------------------------------------------------------- |

> | `hvc1.1.6.L90.B0`  | 3.0   | 1080p          | [73.6%](https://webcodecsfundamentals.org/codecs/hvc1.1.6.L90.B0.html)  |

Reviewer 4: The HEVC table claims hvc1.1.6.L90.B0 (Level 3.0) supports up to 1080p. In the HEVC spec, Level 3.0 tops out at 960x540 (qHD). You need at least Level 4.0 (L120) to handle 1080p video.

Author: I fixed this with the right level

### Codec-container compatibility

Not all codecs are supported by all containers.
The following table covers the two most common web video containers:

| Codec | MP4     | WebM |
| ----- | ------- | ---- |
| H.264 | Yes     | No   |
| VP9   | Partial | Yes  |
| AV1   | Partial | Yes  |
| HEVC  | Yes     | No   |

H.264 is the standard codec for MP4. VP9 and AV1 are the standard codecs for WebM.
While VP9 and AV1 have partial MP4 support in some environments, pairing them with WebM is more reliable.

### Codec string selection

For each codec family, there are hundreds of possible codec strings.

Each codec string encodes a **profile** and **level** that determine the capabilities and compatibility of the encoded stream. The profile controls which encoding features are enabled — lower profiles such as Baseline are simpler and more broadly compatible, while higher profiles such as High enable better compression at the cost of requiring more capable hardware. The level sets the maximum resolution and bitrate the stream can use. In general, prefer lower profiles and levels unless you specifically need the higher resolution or compression efficiency.

> The following tables provide a practical starting point for codec strings, with levels and profiles that maximize encoding compatibility.

Reviewer 2: Maybe. I wasn't sure how much profile and level affect selection, or that is is just the support number and max resolution that affects your choices. But anyhow.

```suggestion
The following tables provide a practical starting point for selecting your codec strings.
They have been chosen because they offer have broad support across browsers and platforms.
```

Reviewer 2: @sb2702 Ping on this one.

Use {{domxref("AudioEncoder/isConfigSupported_static", "AudioEncoder.isConfigSupported()")}} to check support at runtime before configuring an `AudioEncoder`. Note that `AudioEncoder` itself is not available in all browsers — check for its existence with `typeof AudioEncoder !== "undefined"` before calling `isConfigSupported()`.

> ## Common use cases

Reviewer 2: FYI, Nice!
